projects/driver_controls/src/stm32f0xx/ads1015.c

```diff
@@ -0,0 +1,201 @@
+  }
+  // Update so that the ADS1015 reads from the next channel
+  do {
+    current_channel = (current_channel + 1) % NUM_ADS1015_CHANNELS;
+  } while (!data->channel_enable[current_channel]);
+  prv_set_channel(data, current_channel);
+  data->current_channel = current_channel;
+}
+
+// initiates ads1015 by setting up registers and enabling ALRT/RDY Pin
+StatusCode ads1015_init(ADS1015Data *data, I2CPort i2c_port, ADS1015Address i2c_addr,
+                        GPIOAddress *ready_pin) {
+  if (!data || !ready_pin) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+  data->i2c_port = i2c_port;
+  data->i2c_addr = i2c_addr;
+  data->ready_pin = ready_pin;
+  for (uint8_t channel = 0; channel < NUM_ADS1015_CHANNELS; channel++) {
+    data->channel_enable[channel] = false;
+    data->channel_readings[channel][0] = 0;
+    data->channel_readings[channel][1] = 0;
+    data->channel_callback[channel] = NULL;
+    data->callback_context[channel] = NULL;
+  }
+
+  status_ok_or_return(prv_reset(data));
+  status_ok_or_return(
+      prv_setup_register(data, ADS1015_ADDRESS_POINTER_CONFIG, CONFIG_REGISTER_MSB_0,
+                         CONFIG_REGISTER_LSB));  // sets the starting channel to channel 0
+  status_ok_or_return(prv_setup_register(data, ADS1015_ADDRESS_POINTER_LO_THRESH,
+                                         ADS1015_LO_THRESH_REGISTER_MSB,
+                                         ADS1015_LO_THRESH_REGISTER_LSB));
+  status_ok_or_return(prv_setup_register(data, ADS1015_ADDRESS_POINTER_HI_THRESH,
+                                         ADS1015_HI_THRESH_REGISTER_MSB,
+                                         ADS1015_HI_THRESH_REGISTER_LSB));
+
+  // ALERT/RDY continues to require a pullup resistor.
+  GPIOSettings gpio_settings = {
+    .direction = GPIO_DIR_IN,  //
+  };
+  status_ok_or_return(gpio_init_pin(data->ready_pin, &gpio_settings));
+  return STATUS_CODE_OK;
+}
+
+// This function enable/disables channels, and registers callbacks for each channel.
+// It also registers the interrupt handler on ALRT/RDY pin.
+StatusCode ads1015_configure_channel(ADS1015Data *data, ADS1015Channel channel, bool enable,
+                                     ADS1015Callback callback, void *context) {
+  if (!data) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+  // if no channels were already enabled, set to given channel
+  if (!data->channel_enable[0] && !data->channel_enable[1] && !data->channel_enable[2] &&
+      !data->channel_enable[3]) {
+    status_ok_or_return(prv_set_channel(data, channel));
+    data->current_channel = channel;
+  }
+
+  data->channel_enable[channel] = enable;
+  data->channel_callback[channel] = callback;
+  data->callback_context[channel] = context;
+
+  InterruptSettings it_settings = {
+    .type = INTERRUPT_TYPE_INTERRUPT,
+    .priority = INTERRUPT_PRIORITY_NORMAL,
+  };
+  StatusCode code = gpio_it_register_interrupt(data->ready_pin, &it_settings, INTERRUPT_EDGE_RISING,
```

Comment:
Why do you initialize the interrupt for every single channel?

Reply:
Ah, I did that because I thought it would be necessary as storage gets updated. But since *storage(void *context) is a pointer, registering the interrupt once should be enough.

Reply:
More than that, trying to re-register the interrupt would overwrite the previously registered interrupt handler.